### PR TITLE
Ir/wallet network fix

### PR DESCRIPTION
**Pull request by ioanaroceanu2000**
* Catch this error and only show `wrong network` instead of the raw message 
Screenshot 2022-10-24 at 17 45 56
* Also do live UI updates on network/account change with Metamask

### CODE REVIEW

**Comment by Sol3iman** [comment]
Can you explain the steps involved to reproduce this error? Should I simply connect with the right network and then while connected, switch to the wrong network to test the happy path? @ioanaroceanu2000

File: src/App.tsx [hunk 1]
```diff
@@ -5,6 +5,8 @@ import { routes, TradingLeague, Profile, LiquidityProvider, Trader, FixedBorrowe
 import { AlphaBanner, GweiBar } from '@components/composite';
 import Box from '@mui/material/Box';
 import { useEffect } from 'react';
+import useWallet from './hooks/useWallet';
```
Comment (ssbarbee) [inline src/App.tsx:8]:
```suggestion
import { useWallet } from '@hooks';
```

lets keep same style throughout the code

File: src/contexts/WalletContext/services.ts [hunk 2]
```diff
@@ -18,14 +18,18 @@ const unavailableText = 'Service unavailable, please try again shortly';
...
+  try {
+    const network = await provider.getNetwork();
+    if (network.name !== process.env.REACT_APP_REQUIRED_ETHEREUM_NETWORK) {
+      // eslint-disable-next-line
+      console.warn(
+        `User wallet is connected to '${network.name}' instead of '${
+          process.env.REACT_APP_REQUIRED_ETHEREUM_NETWORK || '<unknown>'
+        }'`,
+      );
+      throw new Error();
```
Comment (ssbarbee) [inline src/contexts/WalletContext/services.ts:30]:
```suggestion
      throw new Error(`User wallet is connected to '${network.name}' instead of '${
          process.env.REACT_APP_REQUIRED_ETHEREUM_NETWORK || '<unknown>'
        }'`);
```

remove not needed console.warn and eslint-disable-next-line... Enrich the Error you throw

File: src/types/ethereum.d.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/// <reference types="react-scripts" />
+
+interface Window {
```
Comment (ssbarbee) [inline src/types/ethereum.d.ts:4]: Why is this needed? Please update the already existing global.d.ts file, with the new methods to added in App.tsx

File: src/App.tsx [hunk 4]
```diff
@@ -13,6 +15,22 @@ const App = () => {
     }
   }, []);
 
+  const wallet = useWallet();
+
+  /* eslint-disable @typescript-eslint/no-unsafe-call */
```
Comment (ssbarbee) [inline src/App.tsx:20]: why eslint-disable?

please update type definitions in global.d.ts

File: src/App.tsx [hunk 5]
```diff
@@ -13,6 +15,22 @@ const App = () => {
...
+  /* eslint-disable @typescript-eslint/no-unsafe-member-access */
+  window.ethereum?.on('accountsChanged', function () {
+    if (!wallet.loading) {
+      void wallet.connect('metamask');
```
Comment (ssbarbee) [inline src/App.tsx:24]: can we actually connect with different wallet other than metamask? Why do we hardcode metamask here? Please explain 🙏

**Comment by ioanaroceanu2000** [comment]
> Can you explain the steps involved to reproduce this error? Should I simply connect with the right network and then while connected, switch to the wrong network to test the happy path? @ioanaroceanu2000

@Sol3iman it's very difficult to test, it occurs randomly. To force it, you need to specify the network in the provider instantiation and then try connecting with another network.

File: src/App.tsx (hunk 5 above)
Comment (ioanaroceanu2000) [reply src/App.tsx:24]: `window.ethereum?.on` listens to metamask events only. So it means a change happened while connected to metamask. We use WalletConnect as well but the WalletConnect flow is not affected.

File: src/App.tsx (hunk 1 above)
Comment (ioanaroceanu2000) [reply src/App.tsx]: done in ff692256e500f567310f338bb4a684c73a01433a

File: src/contexts/WalletContext/services.ts (hunk 2 above)
Comment (ioanaroceanu2000) [reply src/contexts/WalletContext/services.ts]: done in ff692256e500f567310f338bb4a684c73a01433a

File: src/types/ethereum.d.ts (hunk 3 above)
Comment (ioanaroceanu2000) [reply src/types/ethereum.d.ts:4]: I added a TODO for metamask 2.0.0. This will solve the issue but let's not risk it before release

**Review by ssbarbee — APPROVED** [review]
LGTM
